File: javascript/radixSort.js
/**
 * Radix Sort (LSD - Least Significant Digit)
 * 
 * Description: Non-comparison sorting algorithm that sorts integers by processing
 * individual digits. Uses counting sort as a subroutine for each digit position.
 * 
 * Time Complexity: O(d × (n + k)) where d is number of digits, k is base
 * Space Complexity: O(n + k)
 * 
 * Use Cases:
 * - Sorting large integers efficiently
 * - Sorting strings of equal length
 * - External sorting with limited memory
 * - Parallel sorting algorithms
 * 
 * Example:
 * Input: arr = [170, 45, 75, 90, 802, 24, 2, 66]
 * Output: [2, 24, 45, 66, 75, 90, 170, 802]
 */

function radixSort(arr, base = 10) {
    // Input validation
    if (!Array.isArray(arr) || arr.length === 0) {
        return [];
    }

    // Handle negative numbers
    const hasNegative = arr.some(num => num < 0);
    if (hasNegative) {
        return radixSortWithNegatives(arr, base);
    }

    const max = Math.max(...arr);
    const maxDigits = Math.floor(Math.log(max) / Math.log(base)) + 1;

    let result = [...arr];

    for (let digit = 0; digit < maxDigits; digit++) {
        result = countingSortByDigit(result, digit, base);
    }

    return result;
}

function countingSortByDigit(arr, digitPos, base) {
    const n = arr.length;
    const output = new Array(n);
    const count = new Array(base).fill(0);

    // Get digit at position
    const getDigit = (num, pos, base) => {
        return Math.floor(num / Math.pow(base, pos)) % base;
    };

    // Count occurrences
    for (let num of arr) {
        const digit = getDigit(num, digitPos, base);
        count[digit]++;
    }

    // Cumulative count
    for (let i = 1; i < base; i++) {
        count[i] += count[i - 1];
    }

    // Build output array (stable sort)
    for (let i = n - 1; i >= 0; i--) {
        const digit = getDigit(arr[i], digitPos, base);
        output[count[digit] - 1] = arr[i];
        count[digit]--;
    }

    return output;
}

// Radix sort with negative numbers
function radixSortWithNegatives(arr, base = 10) {
    const negative = arr.filter(x => x < 0).map(x => -x);
    const positive = arr.filter(x => x >= 0);

    const sortedNegative = radixSort(negative, base).reverse().map(x => -x);
    const sortedPositive = radixSort(positive, base);

    return [...sortedNegative, ...sortedPositive];
}

// MSD (Most Significant Digit) Radix Sort
function radixSortMSD(arr, base = 10) {
    if (arr.length <= 1) return arr;

    const max = Math.max(...arr.map(Math.abs));
    const maxDigits = Math.floor(Math.log(max) / Math.log(base)) + 1;

    function msdSort(arr, digit, base) {
        if (arr.length <= 1 || digit < 0) return arr;

        const buckets = Array.from({ length: base }, () => []);
        
        for (let num of arr) {
            const d = Math.floor(num / Math.pow(base, digit)) % base;
            buckets[d].push(num);
        }

        const result = [];
        for (let bucket of buckets) {
            result.push(...msdSort(bucket, digit - 1, base));
        }

        return result;
    }

    return msdSort(arr, maxDigits - 1, base);
}

// Radix sort for strings
function radixSortStrings(arr) {
    if (arr.length === 0) return [];

    const maxLen = Math.max(...arr.map(s => s.length));
    let result = [...arr];

    // Sort from rightmost character to leftmost
    for (let pos = maxLen - 1; pos >= 0; pos--) {
        result = countingSortStringsByChar(result, pos);
    }

    return result;
}

function countingSortStringsByChar(arr, charPos) {
    const buckets = Array.from({ length: 256 }, () => []);

    for (let str of arr) {
        const charCode = charPos < str.length ? str.charCodeAt(charPos) : 0;
        buckets[charCode].push(str);
    }

    return buckets.flat();
}

// Radix sort for floating point numbers
function radixSortFloat(arr, precision = 2) {
    const multiplier = Math.pow(10, precision);
    const integers = arr.map(x => Math.round(x * multiplier));
    const sorted = radixSortWithNegatives(integers);
    return sorted.map(x => x / multiplier);
}

// In-place radix sort (uses less memory)
function radixSortInPlace(arr, base = 10) {
    const max = Math.max(...arr);
    const maxDigits = Math.floor(Math.log(max) / Math.log(base)) + 1;

    for (let digit = 0; digit < maxDigits; digit++) {
        const buckets = Array.from({ length: base }, () => []);
        
        for (let num of arr) {
            const d = Math.floor(num / Math.pow(base, digit)) % base;
            buckets[d].push(num);
        }

        let index = 0;
        for (let bucket of buckets) {
            for (let num of bucket) {
                arr[index++] = num;
            }
        }
    }

    return arr;
}

// Radix sort with different bases comparison
function compareRadixBases(arr) {
    const bases = [2, 8, 10, 16];
    const results = {};

    for (let base of bases) {
        const start = performance.now();
        radixSort([...arr], base);
        const end = performance.now();
        results[`base-${base}`] = (end - start).toFixed(3) + 'ms';
    }

    return results;
}

// Count sort for specific digit (helper)
function getMaxDigits(arr, base) {
    const max = Math.max(...arr);
    return Math.floor(Math.log(max) / Math.log(base)) + 1;
}

// Test cases
const arr = [170, 45, 75, 90, 802, 24, 2, 66];

console.log('LSD Radix Sort:', radixSort(arr)); // [2, 24, 45, 66, 75, 90, 170, 802]
console.log('MSD Radix Sort:', radixSortMSD(arr)); // [2, 24, 45, 66, 75, 90, 170, 802]

const withNegatives = [170, -45, 75, -90, 802, 24, -2, 66];
console.log('With negatives:', radixSortWithNegatives(withNegatives));

const strings = ['apple', 'pie', 'about', 'cat', 'zoo', 'able'];
console.log('String sort:', radixSortStrings(strings));

const floats = [3.14, 2.71, 1.41, 2.23, 3.16];
console.log('Float sort:', radixSortFloat(floats));

console.log('Base comparison:', compareRadixBases(arr));

module.exports = { 
    radixSort, 
    radixSortMSD, 
    radixSortWithNegatives,
    radixSortStrings,
    radixSortFloat,
    radixSortInPlace
};

File: python/radix_sort.py
# Radix Sort Algorithm
# Time Complexity: O(d * (n + k)) where d is number of digits, k is range of digits
# Space Complexity: O(n + k)
# Stable sorting algorithm

def counting_sort_by_digit(arr, exp):
    """
    A function to do counting sort of arr[] according to the digit represented by exp.
    """
    n = len(arr)
    output = [0] * n  # output array
    count = [0] * 10  # count array for digits 0-9
    
    # Store count of occurrences in count[]
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1
    
    # Change count[i] so that count[i] now contains actual position of this digit in output[]
    for i in range(1, 10):
        count[i] += count[i - 1]
    
    # Build the output array
    i = n - 1
    while i >= 0:
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1
    
    # Copy the output array to arr[], so that arr[] now contains sorted numbers
    for i in range(n):
        arr[i] = output[i]

def radix_sort(arr):
    """
    Main function to implement radix sort.
    """
    # Find the maximum number to know number of digits
    max_num = max(arr)
    
    # Do counting sort for every digit
    exp = 1
    while max_num // exp > 0:
        counting_sort_by_digit(arr, exp)
        exp *= 10

def radix_sort_with_steps(arr):
    """
    Radix sort with step-by-step visualization.
    """
    print(f"Original array: {arr}")
    max_num = max(arr)
    print(f"Maximum number: {max_num}")
    
    exp = 1
    step = 1
    while max_num // exp > 0:
        print(f"\nStep {step}: Sorting by digit at position {exp}")
        print(f"Before sorting by digit: {arr}")
        counting_sort_by_digit(arr, exp)
        print(f"After sorting by digit: {arr}")
        exp *= 10
        step += 1
    
    return arr

def radix_sort_strings(arr):
    """
    Radix sort for strings (lexicographic order).
    """
    if not arr:
        return arr
    
    # Find the maximum length
    max_len = max(len(s) for s in arr)
    
    # Pad strings with spaces to make them same length
    padded_arr = [s.ljust(max_len) for s in arr]
    
    # Sort by each character position from right to left
    for pos in range(max_len - 1, -1, -1):
        # Use counting sort for this position
        count = [0] * 256  # ASCII range
        output = [''] * len(padded_arr)
        
        # Count occurrences
        for s in padded_arr:
            count[ord(s[pos])] += 1
        
        # Calculate positions
        for i in range(1, 256):
            count[i] += count[i - 1]
        
        # Build output array
        for i in range(len(padded_arr) - 1, -1, -1):
            output[count[ord(padded_arr[i][pos])] - 1] = padded_arr[i]
            count[ord(padded_arr[i][pos])] -= 1
        
        padded_arr = output
    
    # Remove padding and return
    return [s.rstrip() for s in padded_arr]

# Example usage
if __name__ == "__main__":
    # Test case 1 - integers
    print("Radix Sort for Integers:")
    arr1 = [170, 45, 75, 90, 2, 802, 24, 66]
    print(f"Original: {arr1}")
    radix_sort(arr1)
    print(f"Sorted: {arr1}")
    
    # Test case 2 with steps
    print("\n" + "="*50)
    print("Radix Sort with Steps:")
    arr2 = [329, 457, 657, 839, 436, 720, 355]
    radix_sort_with_steps(arr2)
    print(f"Final result: {arr2}")
    
    # Test case 3 - strings
    print("\n" + "="*50)
    print("Radix Sort for Strings:")
    words = ["word", "category", "apple", "banana", "cat", "dog"]
    print(f"Original: {words}")
    sorted_words = radix_sort_strings(words)
    print(f"Sorted: {sorted_words}")
    
    # Test case 4 - negative numbers (requires modification)
    print("\n" + "="*50)
    print("Note: This implementation doesn't handle negative numbers.")
    print("For negative numbers, you would need to separate positive and negative arrays.")
